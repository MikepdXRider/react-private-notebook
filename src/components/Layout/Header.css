.header {
  display: grid;
  grid: 50px / 1fr 0.75fr;
  align-items: center;
  justify-items: space-around;
  border-bottom: 2px dashed var(--gray);
  padding: 10px;
}

.header a {
  text-decoration: none;
}

.nav {
  background-color: var(--gray);
  padding: 10px;
}

.nav a,
.nav span {
  color: var(--off-white);
}

.signInStatus {
  display: flex;
  align-items: center;
  justify-content: flex-end;
  gap: 1rem;
}
