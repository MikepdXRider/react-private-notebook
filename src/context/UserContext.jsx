import { createContext, useContext, useMemo, useState } from 'react';
import { getUser } from '../services/users';

const UserContext = createContext();

const UserProvider = ({ children }) => {
  // Is this just for initializing? 
  // OR... When setUser is called, is currentUser called again? 
  const currentUser = getUser();
  console.log(currentUser);
  const [user, setUser] = useState(
    currentUser ? { id: currentUser.id, email: currentUser.email } : {}
    );
  console.log(user);

  const value = useMemo(() => ({ user, setUser }), [user]);

  return <UserContext.Provider value={value}>{children}</UserContext.Provider>;
};

const useUser = () => {
  const context = useContext(UserContext);

  if (context === undefined) {
    throw new Error('useUser must be used within a UserProvider');
  }

  return context;
};

export { UserContext, UserProvider, useUser };
