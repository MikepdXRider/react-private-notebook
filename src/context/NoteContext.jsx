import {
  createContext,
  useContext,
  useEffect,
  useReducer,
  useState,
} from 'react';
import {
  createNote,
  deleteNoteById,
  getNotes,
  updateNoteById,
} from '../services/notes';
import { parseNote, parseNotes } from '../utils/noteParser';
import { useUser } from './UserContext';

// payload is a note object:
// { title: String, content: String }
function notesReducer(notes, { type, payload }) {
  switch (type) {
    case 'create':
      return [payload, ...notes];
    case 'reset':
      return [...payload];
    case 'update':
      return notes.map((note) => (note.id === payload.id ? payload : note));
    case 'delete':
      return notes.filter((note) => note.id !== payload.id);
    default:
      throw Error(`Unknown action: ${type}`);
  }
}

export const NoteContext = createContext();

const NoteProvider = ({ children }) => {
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState('');
  const [notesStore, dispatch] = useReducer(notesReducer, []);
  const { user } = useUser();

  useEffect(() => {
    const fetchNotes = async () => {
      if (user.id) {
        const notes = await getNotes();
        dispatch({
          type: 'reset',
          payload: parseNotes(notes),
        });
        setLoading(false);
      }
    };
    fetchNotes();
  }, [user?.id]);

  const addNote = async ({ title, content }) => {
    setLoading(true);
    try {
      const [note] = await createNote({ userId: user.id, title, content });
      const payload = parseNote(note);
      dispatch({ type: 'create', payload });
      return payload;
    } catch (error) {
      setError(error.message);
    } finally {
      setLoading(false);
    }
  };

  const getNote = (id) => {
    return notesStore.find((note) => note.id === Number(id));
  };

  const updateNote = async ({ id, title, content }) => {
    setLoading(true);
    try {
      const [note] = await updateNoteById(id, { title, content });
      const payload = parseNote(note);
      dispatch({ type: 'update', payload });
      return parseNote(payload);
    } catch (error) {
      setError(error.message);
    } finally {
      setLoading(false);
    }
  };

  const deleteNote = async (id) => {
    setLoading(true);
    try {
      const [note] = await deleteNoteById(id);
      dispatch({ type: 'delete', payload: { id: note.id } });
      return parseNote(note);
    } catch (error) {
      setError(error.message);
    } finally {
      setLoading(false);
    }
  };

  return (
    <NoteContext.Provider
      value={{
        notes: { loading, error, data: notesStore },
        addNote,
        getNote,
        updateNote,
        deleteNote,
      }}
    >
      {children}
    </NoteContext.Provider>
  );
};

const useNotes = () => {
  const context = useContext(NoteContext);

  if (context === undefined) {
    throw new Error('useNotes must be used within a NoteProvider');
  }

  return context;
};

export { NoteProvider, useNotes };
