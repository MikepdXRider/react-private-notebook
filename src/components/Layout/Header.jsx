import { Link } from 'react-router-dom';
import { useUser } from '../../context/UserContext';
import AuthButton from '../AuthButton/AuthButton';
import Breadcrumbs from '../Breadcrumbs/Breadcrumbs';
import styles from './Header.css';

export default function Header() {
  const { user } = useUser();

  return (
    <>
      <header className={styles.header}>
        <Link to="/">
          <h1>My Notes</h1>
        </Link>
        <p className={styles.signInStatus}>
          {user?.email ? `Signed in as ${user?.email}` : 'Not Signed In'}
          <AuthButton />
        </p>
      </header>
      <Breadcrumbs className={styles.nav} />
    </>
  );
}
