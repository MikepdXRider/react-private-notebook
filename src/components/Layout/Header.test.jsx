import { render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { NoteProvider } from '../../context/NoteContext';
import { UserProvider } from '../../context/UserContext';
import Header from './Header';

jest.mock('../../context/UserContext');
jest.mock('../../context/NoteContext');
jest.mock('../../services/users');

it('should display a header with breadcrumbs when signed out', () => {
  const { container } = render(
    <UserProvider>
      <NoteProvider>
        <MemoryRouter>
          <Header />
        </MemoryRouter>
      </NoteProvider>
    </UserProvider>
  );

  expect(container).toMatchSnapshot();
});

it('should display the userâ€™s email when signed in', () => {
  const { container } = render(
    <UserProvider mockUser={{ id: 1, email: 'test@example.com' }}>
      <NoteProvider>
        <MemoryRouter>
          <Header />
        </MemoryRouter>
      </NoteProvider>
    </UserProvider>
  );
  screen.getByText('test@example.com', { exact: false });
  expect(container).toMatchSnapshot();
});
