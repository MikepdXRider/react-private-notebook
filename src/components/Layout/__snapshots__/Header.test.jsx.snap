// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display a header with breadcrumbs when signed out 1`] = `
<div>
  <header
    class="header"
  >
    <a
      href="/"
    >
      <h1>
        My Notes
      </h1>
    </a>
    <p
      class="signInStatus"
    >
      Not Signed In
      <a
        href="/login"
      >
        <button>
          Sign In
        </button>
      </a>
    </p>
  </header>
  <nav
    class="nav"
  >
    <a
      href="/notes"
    >
      Notebook
    </a>
  </nav>
</div>
`;

exports[`should display the userâ€™s email when signed in 1`] = `
<div>
  <header
    class="header"
  >
    <a
      href="/"
    >
      <h1>
        My Notes
      </h1>
    </a>
    <p
      class="signInStatus"
    >
      Signed in as test@example.com
      <button>
        Sign Out
      </button>
    </p>
  </header>
  <nav
    class="nav"
  >
    <a
      href="/notes"
    >
      Notebook
    </a>
  </nav>
</div>
`;
